File: Week_02/PPMH_BT6/Table 8/PPMH_BT6_Table08.py
def implies(a: bool, b: bool) -> bool:
    return (not a) or b

def equiv(a: bool, b: bool) -> bool:
    return a == b

def biconditional(a: bool, b: bool) -> bool:
    return a == b

def main():
    ok = True

    # p <=> q equiv (p => q) ∧ (q => p)
    ok = True
    for i in range(4):
        p = bool(i // 2)
        q = bool(i % 2)
        left = biconditional(p, q)
        right = implies(p, q) and implies(q, p)
        if not equiv(left, right):
            ok = False
    print("p <=> q equiv (p => q) ^ (q => p):", "True" if ok else "False")

    # p <=> q equiv !p <=> !q
    ok = True
    for i in range(4):
        p = bool(i // 2)
        q = bool(i % 2)
        left = biconditional(p, q)
        right = biconditional(not p, not q)
        if not equiv(left, right):
            ok = False
    print("p <=> q equiv !p <=> !q:", "True" if ok else "False")

    # p <=> q equiv (p ∧ q) ∨ (!p ∧ !q)
    ok = True
    for i in range(4):
        p = bool(i // 2)
        q = bool(i % 2)
        left = biconditional(p, q)
        right = (p and q) or ((not p) and (not q))
        if not equiv(left, right):
            ok = False
    print("p <=> q equiv (p ^ q) V (!p ^ !q):", "True" if ok else "False")

    # !(p <=> q) equiv p <=> !q
    ok = True
    for i in range(4):
        p = bool(i // 2)
        q = bool(i % 2)
        left = not biconditional(p, q)
        right = biconditional(p, not q)
        if not equiv(left, right):
            ok = False
    print("!(p <=> q) equiv p <=> !q:", "True" if ok else "False")


if __name__ == "__main__":
    main()

File: Week_02/PPMH_BT6/Table 8/PPMH_BT6_Table08.cpp
#include <iostream>
using namespace std;

bool implies(bool a, bool b){
    return (!a) || b; 
}

bool equiv(bool a, bool b){
    return a == b; 
}

bool biconditional(bool a, bool b){
    return a == b; 
}

int main() {
    bool p, q;
    bool ok;

    // p biconditional q equiv (p => q) ^ (q => p)
    ok = true;
    for(int i = 0; i < 4; i++){
        p = i / 2;
        q = i % 2;
        bool left = biconditional(p, q);
        bool right = implies(p, q) && implies(q, p);
        if (!equiv(left, right)) ok = false;
    }
    cout << "p <=> q equiv (p => q) ^ (q => p): " << (ok ? "True" : "False") << endl;

    // p <=> q equiv !p <=> !q
    ok = true;
    for(int i = 0; i < 4; i++){
        p = i / 2;
        q = i % 2;
        bool left = biconditional(p, q);
        bool right = biconditional(!p, !q);
        if (!equiv(left, right)) ok = false;
    }
    cout << "p <=> q equiv !p <=> !q: " << (ok ? "True" : "False") << endl;

    // p <=> q equiv (p 6 q) V (!p ^ !q)
    ok = true;
    for(int i = 0; i<4; i++){
        p = i / 2;
        q = i % 2;
        bool left = biconditional(p, q);
        bool right = (p && q) || (!p && !q);
        if (!equiv(left, right)) ok = false;
    }
    cout << "p <=> q equiv (p ^ q) V (!p ^ !q): " << (ok ? "True" : "False") << endl;

    // !(p <=> q) equiv p <=> !q
    ok = true;
    for(int i = 0; i < 4; i++){
        p = i / 2;
        q = i % 2;
        bool left = !biconditional(p, q);
        bool right = biconditional(p, !q);
        if (!equiv(left, right)) ok = false;
    }
    cout << "!(p <=> q) equiv p <=> !q: " << (ok ? "True" : "False") << endl;

    return 0;
}